tests/ci/ci_cache.py

```diff
@@ -0,0 +1,818 @@
+            for job in self.jobs_to_wait:
+                if job not in Build:
+                    remove_from_wait.append(job)
+            for job in remove_from_wait:
+                del self.jobs_to_wait[job]
+
+        while (
+            len(self.jobs_to_wait) > MAX_JOB_NUM_TO_WAIT
+            and round_cnt < MAX_ROUNDS_TO_WAIT
+        ):
+            await_finished: Set[str] = set()
+            round_cnt += 1
+            GHActions.print_in_group(
+                f"Wait pending jobs, round [{round_cnt}/{MAX_ROUNDS_TO_WAIT}]:",
+                list(self.jobs_to_wait),
+            )
+            # this is an initial approach to wait pending jobs:
+            # start waiting for the next TIMEOUT seconds if there are more than X(=4) jobs to wait
+            # wait TIMEOUT seconds in rounds. Y(=5) is the max number of rounds
+            expired_sec = 0
+            start_at = int(time.time())
+            while expired_sec < TIMEOUT and self.jobs_to_wait:
+                time.sleep(poll_interval_sec)
+                self.update()
+                for job_name, job_config in self.jobs_to_wait.items():
+                    num_batches = job_config.num_batches
+                    job_config = CI_CONFIG.get_job_config(job_name)
+                    assert job_config.pending_batches
+                    assert job_config.batches
+                    pending_batches = list(job_config.pending_batches)
+                    for batch in pending_batches:
+                        if self.is_pending(
+                            job_name,
+                            batch,
+                            num_batches,
+                            release_branch=is_release
+                            and job_config.required_on_release_branch,
+                        ):
+                            continue
+                        if self.is_successful(
+                            job_name,
+                            batch,
+                            num_batches,
+                            release_branch=is_release
+                            and job_config.required_on_release_branch,
+                        ):
+                            print(
+                                f"Job [{job_name}_[{batch}/{num_batches}]] is not pending anymore"
+                            )
+                            job_config.batches.remove(batch)
+                            job_config.pending_batches.remove(batch)
+                        else:
+                            print(
+                                f"NOTE: Job [{job_name}:{batch}] finished failed - do not add to ready"
+                            )
+                            job_config.pending_batches.remove(batch)
+
+                        if not job_config.pending_batches:
+                            await_finished.add(job_name)
+
+                for job in await_finished:
+                    self.jobs_to_skip.append(job)
+                    del self.jobs_to_wait[job]
```

Comment:
```
Traceback (most recent call last):
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/ci.py", line 1238, in <module>
    sys.exit(main())
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/ci.py", line 979, in main
    ci_cache.await_pending_jobs(pr_info.is_release)
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/ci_cache.py", line 756, in await_pending_jobs
    del self.jobs_to_wait[job]
KeyError: 'package_msan'
```

https://github.com/ClickHouse/ClickHouse/actions/runs/9380634423/job/25828094637?pr=58349

Reply:
@maxknv mind taking a look why RunConfig timed out and fails eventually? https://github.com/ClickHouse/ClickHouse/pull/58349

Reply:
Looks like it had been fixed by https://github.com/ClickHouse/ClickHouse/pull/64825